File: app/src/main/java/com/byd/dashcast/ui/AppListAdapter.java
package com.byd.dashcast.ui;

import com.byd.dashcast.R;
import androidx.recyclerview.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.CompoundButton;
import android.widget.LinearLayout;

import com.byd.dashcast.model.AppInfo;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

@android.annotation.SuppressLint("SetTextI18n")
public class AppListAdapter extends RecyclerView.Adapter<AppListAdapter.ViewHolder> {

    public interface OnSendToDashboardListener {
        void onSendToDashboard(AppInfo app);
        void onSendToMain(AppInfo app);
        void onKillApp(AppInfo app);
        void onToggleFavorite(AppInfo app);
        void onSetAutoLaunch(AppInfo app, boolean enable);
        /** v0.9.72 — long-press now opens an action bottom sheet (favorite/move/kill/auto). */
        void onShowActions(AppInfo app);
    }

    private List<AppInfo> mAllApps = new ArrayList<>();   // full unfiltered list
    private List<AppInfo> mApps    = new ArrayList<>();   // currently displayed (filtered)
    private final OnSendToDashboardListener mListener;
    private String mCurrentPackage = null;
    private String mMainPackage = null;

    /** v0.9.72 — exposed for the long-press bottom sheet. */
    public String getCurrentPackage() { return mCurrentPackage; }
    public String getMainPackage()    { return mMainPackage; }
    private final HashMap<String, Integer> mPackageIndexMap = new HashMap<>();
    private String mCurrentFilter = "";
    private int mCategoryFilter = 0; // 0=all, 1=nav, 2=media

    /** Foreground tint applied to the active row (cluster) — semi-transparent green. */
    private static final int COLOR_FG_ACTIVE  = 0x1A4CAF50;
    /** Foreground tint applied to a row whose app is running on the main display. */
    private static final int COLOR_FG_ON_MAIN = 0x141565C0;
    /** Reusable ConstantState for foreground tints — avoids allocating a new ColorDrawable per bind. */
    private static final android.graphics.drawable.Drawable.ConstantState CS_FG_ACTIVE =
            new android.graphics.drawable.ColorDrawable(COLOR_FG_ACTIVE).getConstantState();
    private static final android.graphics.drawable.Drawable.ConstantState CS_FG_ON_MAIN =
            new android.graphics.drawable.ColorDrawable(COLOR_FG_ON_MAIN).getConstantState();

    private boolean mIsGridMode = false;

    // Resolved once at construction — onBindViewHolder runs on every scroll
    // recycle, so per-bind getString()/getColor() lookups are avoidable work.
    private final String mFavoritePrefix;
    private final String mCategoryNavLabel;
    private final String mCategoryMediaLabel;
    private final int    mColorFavorite;
    private final int    mColorTextPrimary;
    private final int    mColorShortcutBg;
    private final int    mColorShortcutText;

    public AppListAdapter(android.content.Context ctx, OnSendToDashboardListener listener) {
        mListener           = listener;
        mFavoritePrefix     = ctx.getString(R.string.favorite_prefix);
        mCategoryNavLabel   = ctx.getString(R.string.category_navigation);
        mCategoryMediaLabel = ctx.getString(R.string.category_media);
        mColorFavorite      = ctx.getColor(R.color.favorite_gold);
        mColorTextPrimary   = ctx.getColor(R.color.text_primary);
        mColorShortcutBg    = ctx.getColor(R.color.shortcut_btn_bg);
        mColorShortcutText  = ctx.getColor(R.color.shortcut_btn_text);
    }

    @android.annotation.SuppressLint("NotifyDataSetChanged") // full layout swap
    public void setGridMode(boolean isGridMode) {
        if (mIsGridMode != isGridMode) {
            mIsGridMode = isGridMode;
            notifyDataSetChanged();
        }
    }

    public boolean isGridMode() {
        return mIsGridMode;
    }

    public void setApps(List<AppInfo> apps) {
        mAllApps = apps;
        applyFilter(mCurrentFilter);
    }

    /**
     * Filters the displayed list to entries whose name contains {@code query}.
     * Pass an empty string to clear the filter.
     */
    public void filter(String query) {
        mCurrentFilter = query == null ? "" : query;
        applyFilter(mCurrentFilter);
    }

    /**
     * Filters the displayed list by category.
     * @param category 0=all, AppInfo.CATEGORY_NAVIGATION, AppInfo.CATEGORY_MEDIA
     */
    public void filterByCategory(int category) {
        mCategoryFilter = category;
        applyFilter(mCurrentFilter);
    }

    public int getCategoryFilter() {
        return mCategoryFilter;
    }

    @android.annotation.SuppressLint("NotifyDataSetChanged") // bulk filter rebuild
    private void applyFilter(String query) {
        List<AppInfo> base = mAllApps;
        // Category filter
        if (mCategoryFilter != 0) {
            base = new ArrayList<>();
            for (AppInfo a : mAllApps) {
                if (a.category == mCategoryFilter) base.add(a);
            }
        }
        if (query == null || query.trim().isEmpty()) {
            mApps = new ArrayList<>(base);
        } else {
            String lower = query.trim().toLowerCase(java.util.Locale.ROOT);
            List<AppInfo> filtered = new ArrayList<>();
            for (AppInfo a : base) {
                if (a.appName.toLowerCase(java.util.Locale.ROOT).contains(lower)) {
                    filtered.add(a);
                }
            }
            mApps = filtered;
        }
        mPackageIndexMap.clear();
        for (int i = 0; i < mApps.size(); i++) {
            mPackageIndexMap.put(mApps.get(i).packageName, i);
        }
        notifyDataSetChanged();
    }

    public void setCurrentPackage(String packageName) {
        String old = mCurrentPackage;
        mCurrentPackage = packageName;
        notifyPackageChanged(old);
        notifyPackageChanged(packageName);
    }

    public void setMainPackage(String packageName) {
        String old = mMainPackage;
        mMainPackage = packageName;
        notifyPackageChanged(old);
        notifyPackageChanged(packageName);
    }

    private void notifyPackageChanged(String packageName) {
        if (packageName == null) return;
        Integer idx = mPackageIndexMap.get(packageName);
        if (idx != null) notifyItemChanged(idx);
    }

    @Override
    public int getItemViewType(int position) {
        return mIsGridMode ? 1 : 0;
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        int layoutId = viewType == 1 ? R.layout.item_app_grid : R.layout.item_app;
        View v = LayoutInflater.from(parent.getContext()).inflate(layoutId, parent, false);
        return new ViewHolder(v, mListener, this);
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        final AppInfo app = mApps.get(position);
        holder.ivIcon.setImageDrawable(app.icon);
        
        // Indicate pinned state with a star prefix
        if (app.isFavorite) {
            holder.tvName.setText(mFavoritePrefix + app.appName);
            holder.tvName.setTextColor(mColorFavorite);
        } else {
            holder.tvName.setText(app.appName);
            holder.tvName.setTextColor(mColorTextPrimary);
        }

        if (holder.tvCategory != null) {
            if (app.category == AppInfo.CATEGORY_NAVIGATION) {
                holder.tvCategory.setText(mCategoryNavLabel);
                holder.tvCategory.setVisibility(View.VISIBLE);
            } else if (app.category == AppInfo.CATEGORY_MEDIA) {
                holder.tvCategory.setText(mCategoryMediaLabel);
                holder.tvCategory.setVisibility(View.VISIBLE);
            } else {
                holder.tvCategory.setVisibility(View.GONE);
            }
        }

        // Render shortcuts if available. Existing child buttons are reused
        // (retext + relisten) instead of removeAllViews() + new Button per bind:
        // inflating Buttons on every scroll recycle caused allocation and
        // layout-pass churn. Only this adapter ever adds children here.
        if (holder.llShortcutsContainer != null) {
            int want = (app.shortcuts != null) ? app.shortcuts.size() : 0;
            if (want == 0) {
                holder.llShortcutsContainer.setVisibility(View.GONE);
                if (holder.llShortcutsContainer.getChildCount() > 0) {
                    holder.llShortcutsContainer.removeAllViews();
                }
            } else {
                holder.llShortcutsContainer.setVisibility(View.VISIBLE);
                while (holder.llShortcutsContainer.getChildCount() > want) {
                    holder.llShortcutsContainer.removeViewAt(
                            holder.llShortcutsContainer.getChildCount() - 1);
                }
                for (int i = 0; i < want; i++) {
                    final com.byd.dashcast.model.AppShortcut shortcut = app.shortcuts.get(i);
                    Button btn;
                    if (i < holder.llShortcutsContainer.getChildCount()) {
                        btn = (Button) holder.llShortcutsContainer.getChildAt(i);
                    } else {
                        btn = new Button(holder.llShortcutsContainer.getContext());
                        btn.setTextSize(9);
                        btn.setAllCaps(false);
                        btn.setPadding(8, 0, 8, 0);

                        LinearLayout.LayoutParams params = new LinearLayout.LayoutParams(
                            LinearLayout.LayoutParams.WRAP_CONTENT,
                            48
                        );
                        params.setMarginEnd(8);
                        btn.setLayoutParams(params);

                        btn.setBackgroundColor(mColorShortcutBg);
                        btn.setTextColor(mColorShortcutText);

                        holder.llShortcutsContainer.addView(btn);
                    }
                    btn.setText(shortcut.label);
                    btn.setOnClickListener(new View.OnClickListener() {
                        @Override
                        public void onClick(View v) {
                            if (mListener != null) {
                                mListener.onSendToDashboard(app);
                                try {
                                    android.content.pm.LauncherApps la = holder.launcherApps;
                                    if (la != null) {
                                        la.startShortcut(app.packageName, shortcut.id, null, null, android.os.Process.myUserHandle());
                                    }
                                } catch (Exception ignored) {
                                    // Shortcut may have been removed or app uninstalled
                                }
                            }
                        }
                    });
                }
            }
        }
        
        // Handle shortcuts for Grid Mode via PopupMenu
        if (holder.tvBtnShortcuts != null) {
            if (app.shortcuts != null && !app.shortcuts.isEmpty()) {
                holder.tvBtnShortcuts.setVisibility(View.VISIBLE);
                holder.tvBtnShortcuts.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        android.widget.PopupMenu popup = new android.widget.PopupMenu(v.getContext(), v);
                        for (int i = 0; i < app.shortcuts.size(); i++) {
                            popup.getMenu().add(0, i, 0, app.shortcuts.get(i).label);
                        }
                        popup.setOnMenuItemClickListener(new android.widget.PopupMenu.OnMenuItemClickListener() {
                            @Override
                            public boolean onMenuItemClick(android.view.MenuItem item) {
                                com.byd.dashcast.model.AppShortcut chosenShortcut = app.shortcuts.get(item.getItemId());
                                if (holder.btnToCluster != null && holder.btnToCluster.getVisibility() == View.VISIBLE) {
                                    holder.btnToCluster.performClick();
                                } else {
                                    if (mListener != null) {
                                        mListener.onSendToDashboard(app);
                                    }
                                }
                                try {
                                    android.content.pm.LauncherApps la = holder.launcherApps;
                                    if (la != null) {
                                        la.startShortcut(app.packageName, chosenShortcut.id, null, null, android.os.Process.myUserHandle());
                                    }
                                } catch (Exception ignored) {
                                    // Shortcut may have been removed or app uninstalled
                                }
                                return true;
                            }
                        });
                        popup.show();
                    }
                });
            } else {
                holder.tvBtnShortcuts.setVisibility(View.GONE);
            }
        }

        boolean isActive = app.packageName != null && app.packageName.equals(mCurrentPackage);
        boolean isOnMain = app.packageName != null && app.packageName.equals(mMainPackage);
        
        if (holder.viewActiveIndicator != null) {
            holder.viewActiveIndicator.setVisibility((isActive || isOnMain) ? View.VISIBLE : View.GONE);
        }
        // v0.9.72 \u2014 in grid mode keep tiles minimal: icon + name + favorite badge + active dot.
        // All actions (auto-launch, move-to-main/cluster, kill) live in the long-press bottom sheet.
        if (mIsGridMode) {
            if (holder.btnToMain != null) holder.btnToMain.setVisibility(View.GONE);
            if (holder.btnToCluster != null) holder.btnToCluster.setVisibility(View.GONE);
            if (holder.btnKill != null) holder.btnKill.setVisibility(View.GONE);

            if (holder.tvCategory != null) holder.tvCategory.setVisibility(View.GONE);
        } else {
            if (holder.btnToMain != null) {
                holder.btnToMain.setVisibility(isActive ? View.VISIBLE : View.GONE);
            }
            if (holder.btnToCluster != null) {
                holder.btnToCluster.setVisibility(isOnMain ? View.VISIBLE : View.GONE);
            }
            if (holder.btnKill != null) {
                holder.btnKill.setVisibility((isActive || isOnMain) ? View.VISIBLE : View.GONE);
            }
        }

        // Subtle background tint on the active row — preserves the ripple via setForeground().
        // Skipped when the holder already shows the right state: newDrawable()
        // allocates, and at most two rows ever change state per update.
        int fgState = isActive ? 1 : (isOnMain ? 2 : 0);
        if (holder.lastFgState != fgState) {
            holder.lastFgState = fgState;
            if (fgState == 1) {
                holder.itemView.setForeground(CS_FG_ACTIVE.newDrawable());
            } else if (fgState == 2) {
                holder.itemView.setForeground(CS_FG_ON_MAIN.newDrawable());
            } else {
                holder.itemView.setForeground(null);
            }
        }

        // Auto-launch badge on the app icon (list + grid). The toggle itself
        // lives in the long-press bottom sheet (onShowActions → swAuto).
        if (holder.badgeAutoLaunch != null) {
            holder.badgeAutoLaunch.setVisibility(app.isAutoLaunch ? View.VISIBLE : View.GONE);
        }
    }

    @Override
    public int getItemCount() {
        return mApps.size();
    }

    /** Returns the full unfiltered app list (used for auto-launch lookup etc.). */
    public List<AppInfo> getApps() { return mAllApps; }

    AppInfo getAppAt(int position) {
        if (position >= 0 && position < mApps.size()) {
            return mApps.get(position);
        }
        return null;
    }

    static class ViewHolder extends RecyclerView.ViewHolder {
        final ImageView ivIcon;
        final TextView  tvName;
        final TextView  tvCategory;
        final LinearLayout llShortcutsContainer;
        final TextView  tvBtnShortcuts; // Shortcut popup trigger for grid mode
        final View      viewActiveIndicator;
        final Button    btnToMain;
        final Button    btnToCluster;
        final Button    btnKill;
        final TextView  badgeAutoLaunch;
        final android.content.pm.LauncherApps launcherApps;
        /** Foreground tint currently applied to itemView: -1 unknown, 0 none, 1 active, 2 on-main. */
        int lastFgState = -1;

        ViewHolder(View itemView, final OnSendToDashboardListener listener, final AppListAdapter adapter) {
            super(itemView);
            launcherApps = (android.content.pm.LauncherApps)
                    itemView.getContext().getSystemService(android.content.Context.LAUNCHER_APPS_SERVICE);
            ivIcon              = (ImageView) itemView.findViewById(R.id.iv_app_icon);
            tvName              = (TextView)  itemView.findViewById(R.id.tv_app_name);
            tvCategory          = (TextView)  itemView.findViewById(R.id.tv_app_category);
            llShortcutsContainer = itemView.findViewById(R.id.ll_shortcuts_container);
            tvBtnShortcuts      = itemView.findViewById(R.id.tv_btn_shortcuts);
            viewActiveIndicator = itemView.findViewById(R.id.view_active_indicator);
            btnToMain           = (Button)    itemView.findViewById(R.id.btn_to_main);
            btnToCluster        = (Button)    itemView.findViewById(R.id.btn_to_cluster);
            btnKill             = (Button)    itemView.findViewById(R.id.btn_kill_app);
            badgeAutoLaunch     = (TextView)  itemView.findViewById(R.id.badge_auto_launch);

            itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    AppInfo app = adapter.getAppAt(getAdapterPosition());
                    if (app != null && listener != null) listener.onSendToDashboard(app);
                }
            });

            // Long click — v0.9.72 opens the action bottom sheet (was: toggle favorite).
            itemView.setOnLongClickListener(new View.OnLongClickListener() {
                @Override
                public boolean onLongClick(View v) {
                    AppInfo app = adapter.getAppAt(getAdapterPosition());
                    if (app != null && listener != null) {
                        listener.onShowActions(app);
                        return true;
                    }
                    return false;
                }
            });

            if (btnToMain != null) btnToMain.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    AppInfo app = adapter.getAppAt(getAdapterPosition());
                    if (app != null && listener != null) listener.onSendToMain(app);
                }
            });

            if (btnToCluster != null) btnToCluster.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    AppInfo app = adapter.getAppAt(getAdapterPosition());
                    if (app != null && listener != null) listener.onSendToDashboard(app);
                }
            });

            if (btnKill != null) btnKill.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    AppInfo app = adapter.getAppAt(getAdapterPosition());
                    if (app != null && listener != null) listener.onKillApp(app);
                }
            });
        }
    }
}

File: app/src/main/java/com/byd/dashcast/ui/AppListAdapter.kt
package com.byd.dashcast.ui

import android.annotation.SuppressLint
import android.content.Context
import android.content.pm.LauncherApps
import android.graphics.drawable.Drawable
import android.os.Process
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.PopupMenu
import android.widget.TextView
import androidx.core.graphics.drawable.toDrawable
import androidx.core.view.isVisible
import androidx.recyclerview.widget.RecyclerView
import com.byd.dashcast.R
import com.byd.dashcast.model.AppInfo
import java.util.Locale

@SuppressLint("SetTextI18n")
class AppListAdapter(
    ctx: Context,
    private val mListener: OnSendToDashboardListener?
) : RecyclerView.Adapter<AppListAdapter.ViewHolder>() {

    interface OnSendToDashboardListener {
        fun onSendToDashboard(app: AppInfo)
        fun onSendToMain(app: AppInfo)
        fun onKillApp(app: AppInfo)
        fun onToggleFavorite(app: AppInfo)
        fun onSetAutoLaunch(app: AppInfo, enable: Boolean)
        /** v0.9.72 — long-press now opens an action bottom sheet (favorite/move/kill/auto). */
        fun onShowActions(app: AppInfo)
    }

    private var mAllApps: List<AppInfo> = ArrayList() // full unfiltered list
    private var mApps: List<AppInfo> = ArrayList() // currently displayed (filtered)
    private var mCurrentPackage: String? = null
    private var mMainPackage: String? = null
    /** Packages currently projected via a fission layout (headless orchestrator). */
    private var mLayoutPackages: Set<String> = emptySet()

    /** v0.9.72 — exposed for the long-press bottom sheet. */
    fun getCurrentPackage(): String? = mCurrentPackage
    fun getMainPackage(): String? = mMainPackage

    private val mPackageIndexMap = HashMap<String, Int>()
    private var mCurrentFilter = ""
    private var mCategoryFilter = 0 // 0=all, 1=nav, 2=media

    private var mIsGridMode = false

    // Resolved once at construction — onBindViewHolder runs on every scroll recycle.
    private val mFavoritePrefix = ctx.getString(R.string.favorite_prefix)
    private val mCategoryNavLabel = ctx.getString(R.string.category_navigation)
    private val mCategoryMediaLabel = ctx.getString(R.string.category_media)
    private val mColorFavorite = ctx.getColor(R.color.favorite_gold)
    private val mColorTextPrimary = ctx.getColor(R.color.text_primary)
    private val mColorShortcutBg = ctx.getColor(R.color.shortcut_btn_bg)
    private val mColorShortcutText = ctx.getColor(R.color.shortcut_btn_text)

    @SuppressLint("NotifyDataSetChanged") // full layout swap
    fun setGridMode(isGridMode: Boolean) {
        if (mIsGridMode != isGridMode) {
            mIsGridMode = isGridMode
            notifyDataSetChanged()
        }
    }

    fun isGridMode(): Boolean = mIsGridMode

    fun setApps(apps: List<AppInfo>) {
        mAllApps = apps
        applyFilter(mCurrentFilter)
    }

    /** Filters the displayed list to entries whose name contains [query]. */
    fun filter(query: String?) {
        mCurrentFilter = query ?: ""
        applyFilter(mCurrentFilter)
    }

    /** Filters the displayed list by category (0=all, CATEGORY_NAVIGATION, CATEGORY_MEDIA). */
    fun filterByCategory(category: Int) {
        mCategoryFilter = category
        applyFilter(mCurrentFilter)
    }

    fun getCategoryFilter(): Int = mCategoryFilter

    @SuppressLint("NotifyDataSetChanged") // bulk filter rebuild
    private fun applyFilter(query: String?) {
        var base = mAllApps
        // Category filter
        if (mCategoryFilter != 0) {
            base = mAllApps.filter { it.category == mCategoryFilter }
        }
        mApps = if (query == null || query.trim().isEmpty()) {
            ArrayList(base)
        } else {
            val lower = query.trim().lowercase(Locale.ROOT)
            base.filter { it.appName.lowercase(Locale.ROOT).contains(lower) }
        }
        mPackageIndexMap.clear()
        for (i in mApps.indices) {
            mPackageIndexMap[mApps[i].packageName] = i
        }
        notifyDataSetChanged()
    }

    fun setCurrentPackage(packageName: String?) {
        val old = mCurrentPackage
        mCurrentPackage = packageName
        notifyPackageChanged(old)
        notifyPackageChanged(packageName)
    }

    fun setMainPackage(packageName: String?) {
        val old = mMainPackage
        mMainPackage = packageName
        notifyPackageChanged(old)
        notifyPackageChanged(packageName)
    }

    /**
     * Updates the set of packages projected via a fission layout. Refreshes only the
     * rows whose membership changed (union of the previous and new sets).
     */
    fun setLayoutPackages(packages: Set<String>?) {
        val next = if (packages != null) HashSet(packages) else emptySet()
        val changed = HashSet(mLayoutPackages)
        changed.addAll(next)
        mLayoutPackages = next
        for (pkg in changed) notifyPackageChanged(pkg)
    }

    fun getLayoutPackages(): Set<String> = mLayoutPackages

    private fun notifyPackageChanged(packageName: String?) {
        if (packageName == null) return
        val idx = mPackageIndexMap[packageName]
        if (idx != null) notifyItemChanged(idx)
    }

    override fun getItemViewType(position: Int): Int = if (mIsGridMode) 1 else 0

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val layoutId = if (viewType == 1) R.layout.item_app_grid else R.layout.item_app
        val v = LayoutInflater.from(parent.context).inflate(layoutId, parent, false)
        return ViewHolder(v, mListener, this)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val app = mApps[position]
        holder.ivIcon.setImageDrawable(app.icon)

        // Indicate pinned state with a star prefix
        if (app.isFavorite) {
            holder.tvName.text = mFavoritePrefix + app.appName
            holder.tvName.setTextColor(mColorFavorite)
        } else {
            holder.tvName.text = app.appName
            holder.tvName.setTextColor(mColorTextPrimary)
        }

        holder.tvCategory?.let { tvCat ->
            when (app.category) {
                AppInfo.CATEGORY_NAVIGATION -> {
                    tvCat.text = mCategoryNavLabel; tvCat.visibility = View.VISIBLE
                }
                AppInfo.CATEGORY_MEDIA -> {
                    tvCat.text = mCategoryMediaLabel; tvCat.visibility = View.VISIBLE
                }
                else -> tvCat.visibility = View.GONE
            }
        }

        // Render shortcuts if available — existing child buttons are reused.
        holder.llShortcutsContainer?.let { container ->
            val want = app.shortcuts.size
            if (want == 0) {
                container.visibility = View.GONE
                if (container.childCount > 0) container.removeAllViews()
            } else {
                container.visibility = View.VISIBLE
                while (container.childCount > want) {
                    container.removeViewAt(container.childCount - 1)
                }
                for (i in 0 until want) {
                    val shortcut = app.shortcuts[i]
                    val btn: Button
                    if (i < container.childCount) {
                        btn = container.getChildAt(i) as Button
                    } else {
                        btn = Button(container.context)
                        btn.textSize = 9f
                        btn.isAllCaps = false
                        btn.setPadding(8, 0, 8, 0)
                        val params = LinearLayout.LayoutParams(LinearLayout.LayoutParams.WRAP_CONTENT, 48)
                        params.marginEnd = 8
                        btn.layoutParams = params
                        btn.setBackgroundColor(mColorShortcutBg)
                        btn.setTextColor(mColorShortcutText)
                        container.addView(btn)
                    }
                    btn.text = shortcut.label
                    btn.setOnClickListener {
                        if (mListener != null) {
                            mListener.onSendToDashboard(app)
                            try {
                                holder.launcherApps?.startShortcut(
                                    app.packageName, shortcut.id, null, null, Process.myUserHandle()
                                )
                            } catch (ignored: Exception) {
                                // Shortcut may have been removed or app uninstalled
                            }
                        }
                    }
                }
            }
        }

        // Handle shortcuts for Grid Mode via PopupMenu
        holder.tvBtnShortcuts?.let { btnShortcuts ->
            if (app.shortcuts.isNotEmpty()) {
                btnShortcuts.visibility = View.VISIBLE
                btnShortcuts.setOnClickListener { v ->
                    val popup = PopupMenu(v.context, v)
                    for (i in app.shortcuts.indices) {
                        popup.menu.add(0, i, 0, app.shortcuts[i].label)
                    }
                    popup.setOnMenuItemClickListener { item ->
                        val chosenShortcut = app.shortcuts[item.itemId]
                        if (holder.btnToCluster?.isVisible == true) {
                            holder.btnToCluster.performClick()
                        } else {
                            mListener?.onSendToDashboard(app)
                        }
                        try {
                            holder.launcherApps?.startShortcut(
                                app.packageName, chosenShortcut.id, null, null, Process.myUserHandle()
                            )
                        } catch (ignored: Exception) {
                            // Shortcut may have been removed or app uninstalled
                        }
                        true
                    }
                    popup.show()
                }
            } else {
                btnShortcuts.visibility = View.GONE
            }
        }

        val isActive = app.packageName == mCurrentPackage
        val isOnMain = app.packageName == mMainPackage
        // A fission-layout app is shown as "active" and exposes the kill action — but NOT
        // the classic move-to-main/cluster buttons (their ClusterService path is unaware of
        // the fission VD slots).
        val isLayout = mLayoutPackages.contains(app.packageName)

        holder.viewActiveIndicator?.visibility =
            if (isActive || isOnMain || isLayout) View.VISIBLE else View.GONE
        // In grid mode keep tiles minimal; all actions live in the long-press bottom sheet.
        if (mIsGridMode) {
            holder.btnToMain?.visibility = View.GONE
            holder.btnToCluster?.visibility = View.GONE
            holder.btnKill?.visibility = View.GONE
            holder.tvCategory?.visibility = View.GONE
        } else {
            holder.btnToMain?.visibility = if (isActive) View.VISIBLE else View.GONE
            holder.btnToCluster?.visibility = if (isOnMain) View.VISIBLE else View.GONE
            holder.btnKill?.visibility = if (isActive || isOnMain || isLayout) View.VISIBLE else View.GONE
        }

        // Subtle background tint on the active row — preserves the ripple via setForeground().
        val fgState = if (isActive || isLayout) 1 else if (isOnMain) 2 else 0
        if (holder.lastFgState != fgState) {
            holder.lastFgState = fgState
            holder.itemView.foreground = when (fgState) {
                1 -> CS_FG_ACTIVE?.newDrawable()
                2 -> CS_FG_ON_MAIN?.newDrawable()
                else -> null
            }
        }

        // Auto-launch badge on the app icon (list + grid).
        holder.badgeAutoLaunch?.visibility = if (app.isAutoLaunch) View.VISIBLE else View.GONE
    }

    override fun getItemCount(): Int = mApps.size

    /** Returns the full unfiltered app list (used for auto-launch lookup etc.). */
    fun getApps(): List<AppInfo> = mAllApps

    internal fun getAppAt(position: Int): AppInfo? =
        if (position in mApps.indices) mApps[position] else null

    class ViewHolder(
        itemView: View,
        listener: OnSendToDashboardListener?,
        adapter: AppListAdapter
    ) : RecyclerView.ViewHolder(itemView) {
        val ivIcon: ImageView = itemView.findViewById(R.id.iv_app_icon)
        val tvName: TextView = itemView.findViewById(R.id.tv_app_name)
        val tvCategory: TextView? = itemView.findViewById(R.id.tv_app_category)
        val llShortcutsContainer: LinearLayout? = itemView.findViewById(R.id.ll_shortcuts_container)
        val tvBtnShortcuts: TextView? = itemView.findViewById(R.id.tv_btn_shortcuts)
        val viewActiveIndicator: View? = itemView.findViewById(R.id.view_active_indicator)
        val btnToMain: Button? = itemView.findViewById(R.id.btn_to_main)
        val btnToCluster: Button? = itemView.findViewById(R.id.btn_to_cluster)
        val btnKill: Button? = itemView.findViewById(R.id.btn_kill_app)
        val badgeAutoLaunch: TextView? = itemView.findViewById(R.id.badge_auto_launch)
        val launcherApps: LauncherApps? =
            itemView.context.getSystemService(Context.LAUNCHER_APPS_SERVICE) as? LauncherApps
        /** Foreground tint on itemView: -1 unknown, 0 none, 1 active, 2 on-main. */
        var lastFgState = -1

        init {
            itemView.setOnClickListener {
                val app = adapter.getAppAt(adapterPosition)
                if (app != null) listener?.onSendToDashboard(app)
            }
            // Long click — v0.9.72 opens the action bottom sheet.
            itemView.setOnLongClickListener {
                val app = adapter.getAppAt(adapterPosition)
                if (app != null && listener != null) {
                    listener.onShowActions(app)
                    true
                } else {
                    false
                }
            }
            btnToMain?.setOnClickListener {
                val app = adapter.getAppAt(adapterPosition)
                if (app != null) listener?.onSendToMain(app)
            }
            btnToCluster?.setOnClickListener {
                val app = adapter.getAppAt(adapterPosition)
                if (app != null) listener?.onSendToDashboard(app)
            }
            btnKill?.setOnClickListener {
                val app = adapter.getAppAt(adapterPosition)
                if (app != null) listener?.onKillApp(app)
            }
        }
    }

    companion object {
        /** Foreground tint applied to the active row (cluster) — semi-transparent green. */
        private const val COLOR_FG_ACTIVE = 0x1A4CAF50
        /** Foreground tint applied to a row whose app is running on the main display. */
        private const val COLOR_FG_ON_MAIN = 0x141565C0
        // Reusable ConstantState for foreground tints — avoids a new ColorDrawable per bind.
        private val CS_FG_ACTIVE: Drawable.ConstantState? = COLOR_FG_ACTIVE.toDrawable().constantState
        private val CS_FG_ON_MAIN: Drawable.ConstantState? = COLOR_FG_ON_MAIN.toDrawable().constantState
    }
}
